### PR TITLE
optimize: suspend undolog conflict

### PR DESCRIPTION
**Pull request by imherewait**
- [ ] I have registered the PR [changes](https://github.com/seata/seata/tree/develop/changes).

异步删除tcc_fence_log和undo_log中处于悬挂状态的日志

### Ⅱ. Does this pull request fix one issue?
optimization after a unique index conflict occurs in undolog #3160

### CODE REVIEW

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java [hunk 1]
```diff
@@ -241,7 +248,12 @@ public void flushUndoLogs(ConnectionProxy cp) throws SQLException {
             undoLogContent = CompressorFactory.getCompressor(compressorType.getCode()).compress(undoLogContent);
         }
 
-        insertUndoLogWithNormal(xid, branchId, buildContext(parser.getName(), compressorType), undoLogContent, cp.getTargetConnection());
+        try {
+            insertUndoLogWithNormal(xid, branchId, buildContext(parser.getName(), compressorType), undoLogContent, cp.getTargetConnection());
+        }catch (SQLIntegrityConstraintViolationException e){
+            LOGGER.error("Insert undo log duplicate key exception,maybe has been rollbacked by phase2",e);
```
Comment (a364176773) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java:254]: 把异常往外抛出去

Comment (imherewait) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java]: > 把异常往外抛出去

已修改，麻烦有空再帮忙看下

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java [hunk 2]
```diff
@@ -102,6 +107,8 @@ public int getValue() {
 
     private static final ThreadLocal<String> SERIALIZER_LOCAL = new ThreadLocal<>();
 
+    private final AsyncWorker asyncWorker = AsyncWorker.getInstance((DataSourceManager) getResourceManager());
```
Comment (a364176773) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java:110]: private final -> private final

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java [hunk 3]
```diff
@@ -241,7 +248,13 @@ public void flushUndoLogs(ConnectionProxy cp) throws SQLException {
             undoLogContent = CompressorFactory.getCompressor(compressorType.getCode()).compress(undoLogContent);
         }
 
-        insertUndoLogWithNormal(xid, branchId, buildContext(parser.getName(), compressorType), undoLogContent, cp.getTargetConnection());
+        try {
+            insertUndoLogWithNormal(xid, branchId, buildContext(parser.getName(), compressorType), undoLogContent, cp.getTargetConnection());
+        } catch (SQLIntegrityConstraintViolationException sqle) {
+            LOGGER.error("Insert undo log duplicate key exception,maybe has been rollbacked by phase2",sqle);
+            asyncWorker.cleanSuspendUndoLog(xid,branchId,cp.getDataSourceProxy().getResourceId());
+            throw new SQLException(sqle);
```
Comment (a364176773) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java:256]: throw new SQLException(sqle) -> throw sqle

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java (hunk 2 above)
Comment (a364176773) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java:110]: 将DataSourceManager里的AsyncWorker 也改为单例

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java [hunk 4]
```diff
@@ -241,7 +248,13 @@ public void flushUndoLogs(ConnectionProxy cp) throws SQLException {
             undoLogContent = CompressorFactory.getCompressor(compressorType.getCode()).compress(undoLogContent);
         }
 
-        insertUndoLogWithNormal(xid, branchId, buildContext(parser.getName(), compressorType), undoLogContent, cp.getTargetConnection());
+        try {
+            insertUndoLogWithNormal(xid, branchId, buildContext(parser.getName(), compressorType), undoLogContent, cp.getTargetConnection());
+        } catch (SQLIntegrityConstraintViolationException sqle) {
+            LOGGER.error("Insert undo log duplicate key exception,maybe has been rollbacked by phase2",sqle);
+            ASYNC_WORKER.cleanSuspendUndoLog(xid,branchId,cp.getDataSourceProxy().getResourceId());
```
Comment (a364176773) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java:255]: 用p3c的code format把这个pr里新加的代码行都code formating一下,否则ci过不了,自己提交pr前可以mvn clean install -Dcheckstyle=false 跑一下

Comment (imherewait) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java:255]: > 用p3c的code format把这个pr里新加的代码行都code formating一下,否则ci过不了,自己提交pr前可以mvn clean install -Dcheckstyle=false 跑一下

跑过了的哈

**Comment by a364176773** [comment]
image
请检查测试用例

**Comment by imherewait** [comment]
> image 请检查测试用例



> image 请检查测试用例

**Comment by imherewait** [comment]
> image 请检查测试用例

已修复测试用例错误；误点了close pr，麻烦帮忙重新打开

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java (hunk 4 above)
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java:255]: the function is OK, but there are questions in the code design.
RMHandlerAT  --> UndoLogManager
AsyncWorker  --> UndoLogManager

Now a new dependency is added：
UndoLogManager -->  AsyncWorker ?

The code has cyclic dependencies.

Comment (imherewait) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoLogManager.java:255]: > the function is OK, but there are questions in the code design. RMHandlerAT --> UndoLogManager AsyncWorker --> UndoLogManager
> 
> Now a new dependency is added： UndoLogManager --> AsyncWorker ?
> 
> The code has cyclic dependencies.

image
将这段处理逻辑抽出到flushUndoLogs调用方这里是否可以？同时branUndoLog对象的构造也抽到外面去？
image

**Comment by a364176773** [comment]
请解决代码冲突,及ci报错

**Pull request closed without merging**